tests/unit/s2n_self_talk_ktls_test.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_random.h"
+
+/* There are issues with MacOS and FreeBSD so we define the constant ourselves.
+ * https://stackoverflow.com/a/34042435 */
+#define S2N_TEST_INADDR_LOOPBACK 0x7f000001 /* 127.0.0.1 */
+
+/* Unlike our other self-talk tests, this test cannot use AF_UNIX / AF_LOCAL.
+ * For a real self-talk test we need real kernel support for kTLS, and only
+ * AF_INET sockets support kTLS.
+ */
+static S2N_RESULT s2n_new_inet_socket_pair(struct s2n_test_io_pair *io_pair)
+{
+    RESULT_ENSURE_REF(io_pair);
+
+    int listener = socket(AF_INET, SOCK_STREAM, 0);
+    RESULT_ENSURE_GT(listener, 0);
+
+    struct sockaddr_in saddr = { 0 };
+    saddr.sin_family = AF_INET;
+    saddr.sin_addr.s_addr = htonl(S2N_TEST_INADDR_LOOPBACK);
+    saddr.sin_port = 0;
+
+    socklen_t addrlen = sizeof(saddr);
+    RESULT_ENSURE_EQ(bind(listener, (struct sockaddr *) &saddr, addrlen), 0);
+    RESULT_ENSURE_EQ(getsockname(listener, (struct sockaddr *) &saddr, &addrlen), 0);
+    RESULT_ENSURE_EQ(listen(listener, 1), 0);
+
+    io_pair->client = socket(AF_INET, SOCK_STREAM, 0);
+    RESULT_ENSURE_GT(io_pair->client, 0);
+
+    fflush(stdout);
+    pid_t pid = fork();
+    RESULT_ENSURE_GTE(pid, 0);
+    if (pid == 0) {
+        RESULT_ENSURE_EQ(connect(io_pair->client, (struct sockaddr *) &saddr, addrlen), 0);
+        ZERO_TO_DISABLE_DEFER_CLEANUP(io_pair);
+        exit(0);
+    }
+    io_pair->server = accept(listener, NULL, NULL);
+    RESULT_ENSURE_GT(io_pair->server, 0);
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* ktls is complicated to enable. We should ensure that it's actually enabled
+     * where we think we're testing it.
+     */
+    bool ktls_expected = (getenv("S2N_KTLS_TESTING_EXPECTED") != NULL);
```

Comment:
I tried writing it where we check the value with strcmp, but it gets more complicated with no real benefit. It mostly has a value because the buildspec requires a value, not because that value is useful. Because why would you ever set it to "false"? Would we really want to assert that a CI job compiles with ktls support but fails to enable it at runtime?

This is also a very common pattern in our code. We usually call "ifdef /  if defined()" instead of checking that macros == 1.